### PR TITLE
smart-contracts, coverage, and tests

### PR DESCRIPTION
**Pull request by andygray**
* [x] `CudosToken` ERC20 smart contract
* [x] Unit tests
* [x] Coverage plugin 
* [x] Truffle framework config
* [x] yarn and github actions setup (with ganache-cli)
* [x] `Standard` code linting

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,41 @@
...
+  * Defines the total supply of tokens - `10 billion` tokens are created upon contract creation
+* The token has a `transfersEnabled` flag that can be called once and once only to enable transfers for all. Only whitelisted admin addresses can call this function.
+* The token has a `whitelisted` addresses that can transfer tokens even if the `transfersEnabled` flag is set to false. Only whitelisted admin addresses can add other addresses to the whitelist.
+
+
+## Local Installation & Testing
+
+1. Install [Truffle](http://truffleframework.com), [Ganache](https://www.trufflesuite.com/ganache) and [NodeJs](https://nodejs.org/en/)(version 10.x upwards) globally
+```bash
+npm install -g truffle
```
Comment (peterwillis) [inline README.md]: could you convert the project to use yarn instead of npm please?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  "version": "1.0.0",
+  "description": "Cudos Smart Contracts",
+  "directories": {
+    "test": "test"
+  },
+  "scripts": {
+    "test": "truffle test --network development",
+    "coverage": "./node_modules/.bin/solidity-coverage"
+  },
+  "author": "BlockRocket.tech",
```
Comment (peterwillis) [inline package.json]: could you remove the author field please?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  "license": "MIT",
+  "devDependencies": {
```
Comment (peterwillis) [inline package.json]: could you add `standard` as a dev dependency and run `standard --fix` on the project?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,41 @@
...
+```
+	
+2. Install dependencies.
+```bash
+npm install
+```
+
+3. Run tests. 
+```bash
+npm run test
```
Comment (peterwillis) [inline README.md]: could you add github actions to the project so that the unit tests are run on pull request update?

File: truffle-config.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        // live: {
+        //     provider: function () {
+        //         return new HDWalletProvider(MNEMONIC, `https://mainnet.infura.io/v3/${INFURA_KEY}`);
+        //     },
+        //     network_id: 1,
+        //     gas: 6075039,         // default = 4712388
+        //     gasPrice: 4100000000, // default = 100 gwei = 100000000000
+        //     timeoutBlocks: 200,   // # of blocks before a deployment times out  (minimum/default: 50)
+        //     skipDryRun: true      // Skip dry run before migrations? (default: false for public nets )
+        // },
```
Comment (peterwillis) [inline truffle-config.js]: If this is not used, it should be removed.

File: README.md (hunk 1 above)
Comment (andygray) [reply README.md]: Done.

File: package.json (hunk 2 above)
Comment (andygray) [reply package.json]: Done

File: README.md (hunk 4 above)
Comment (andygray) [reply README.md]: Done

File: truffle-config.js (hunk 5 above)
Comment (andygray) [reply truffle-config.js]: Removed.

File: cudos-token.iml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="WEB_MODULE" version="4">
+  <component name="NewModuleRootManager" inherit-compiler-output="true">
+    <exclude-output />
+    <content url="file://$MODULE_DIR$" />
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+</module>
```
Comment (cudobuild) [inline cudos-token.iml]: Is this file for an IDE?

Comment (peterwillis) [inline cudos-token.iml]: Is this file for an IDE?

File: migrations/2_deploy_CudosToken.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+const CudosToken = artifacts.require('CudosToken');
+
+module.exports = async function(deployer, network, accounts) {
+  await deployer.deploy(CudosToken);
```
Comment (peterwillis) [inline migrations/2_deploy_CudosToken.js]: Just a question here - do you specifically want this function to ignore the response from deploy? The way this is written means that the function will return a promise, and the value from deploy is explicitly hidden. The function will return a promise that resolves with an empty object.

File: package.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    "chai": "^4.2.0",
+    "eth-gas-reporter": "^0.2.11",
+    "openzeppelin-test-helpers": "^0.3.2",
+    "solidity-coverage": "github:rotcivegaf/solidity-coverage",
+    "truffle-flattener": "^1.4.2"
+  },
+  "dependencies": {
+    "@truffle/hdwallet-provider": "^1.0.18",
+    "ganache-cli": "^6.7.0",
+    "truffle": "^5.0.37"
```
Comment (peterwillis) [inline package.json]: are these all definitely dependencies, not dev dependencies? Isn't truffle a testing framework?

Comment (andygray) [reply package.json]: added all to dev-dependencies - we do not deploy this anywhere - but you are correct we can add them all as dev dependencies.

File: migrations/2_deploy_CudosToken.js (hunk 7 above)
Comment (andygray) [reply migrations/2_deploy_CudosToken.js]: Adjusted - for single line migrations we do not need the `await` we will when we start whitelisting addresses - this is a hangover from that local testing.

File: cudos-token.iml (hunk 6 above)
Comment (andygray) [reply cudos-token.iml]: Removing. This was git ignored but your fork is a slightly different project name. Using *.iml now

Comment (andygray) [reply cudos-token.iml]: Removed

File: package.json (hunk 8 above)
Comment (peterwillis) [reply package.json]: Even @openzeppelin/contracts? Is that resolved by whatever runs the solidity in a different way not through node?

Comment (andygray) [reply package.json]: Truffle is a Smart-Contract framework with testing and migrations etc. Ganache is local blockchain in JS.

All in Dev Dependencies.

File: migrations/2_deploy_CudosToken.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+const CudosToken = artifacts.require('CudosToken')
+
+module.exports = async function (deployer, network, accounts) {
+  deployer.deploy(CudosToken)
```
Comment (peterwillis) [inline migrations/2_deploy_CudosToken.js]: Unless you specifically need this to return a promise, you can remove `async`

Comment (peterwillis) [reply migrations/2_deploy_CudosToken.js]: @andygray I think this is the only remaining comment

**Pull request merged**